crates/rpc/rpc/src/eth/filter.rs

```diff
@@ -553,63 +557,79 @@ where
     /// Returns an error if:
     ///  - underlying database error
     async fn get_logs_in_block_range_inner(
-        &self,
+        self: Arc<Self>,
         filter: &Filter,
         from_block: u64,
         to_block: u64,
         limits: QueryLimits,
     ) -> Result<Vec<Log>, EthFilterError> {
         let mut all_logs = Vec::new();
+        let mut matching_headers = Vec::new();
 
-        // loop over the range of new blocks and check logs if the filter matches the log's bloom
-        // filter
+        // get current chain tip to determine processing mode
+        let chain_tip = self.provider().best_block_number()?;
+
+        // first collect all headers that match the bloom filter for cached mode decision
         for (from, to) in
             BlockRangeInclusiveIter::new(from_block..=to_block, self.max_headers_range)
         {
             let headers = self.provider().headers_range(from..=to)?;
-            for (idx, header) in headers
-                .iter()
-                .enumerate()
-                .filter(|(_, header)| filter.matches_bloom(header.logs_bloom()))
-            {
-                // these are consecutive headers, so we can use the parent hash of the next
-                // block to get the current header's hash
-                let block_hash = match headers.get(idx + 1) {
-                    Some(child) => child.parent_hash(),
+
+            let mut headers_iter = headers
+                .into_iter()
+                .filter(|header| filter.matches_bloom(header.logs_bloom()))
+                .peekable();
+
+            while let Some(header) = headers_iter.next() {
+                let block_hash = match headers_iter.peek() {
+                    Some(child_header) => child_header.parent_hash(),
                     None => self
                         .provider()
                         .block_hash(header.number())?
                         .ok_or_else(|| ProviderError::HeaderNotFound(header.number().into()))?,
                 };
```

Comment:
let's actually replace this case with just seal_slow, I think this a) cleaner and b) probably even faster

Reply:
yep, done